### PR TITLE
Add Siddhi Apps REST Api

### PR DESCRIPTION
**Pull request by LakshanSS**
>wso2.dashboard.datasearch:
>&nbsp;&nbsp;&nbsp;workerAccessCredentials:
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;username: admin
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;password: admin
>&nbsp;&nbsp;&nbsp;workerList:
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- localhost:9443
>     

>Send GET request to https://localhost:9643/portal/apis/datasearch/siddhi-apps to get siddhi apps

>Send GET request to https://localhost:9643/portal/apis/datasearch/worker_id/siddhiAppName
>Ex: https://localhost:9643/portal/apis/datasearch/localhost_9443/ShipmentHistoryApp

### CODE REVIEW

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/ApiResponseMessage.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        date = "2018-08-01T11:26:25.925Z")
+
+public class ApiResponseMessage {
+    public static final int ERROR = 1;
+    public static final int WARNING = 2;
+    public static final int INFO = 3;
+    public static final int OK = 4;
+    public static final int TOO_BUSY = 5;
+
+    int code;
```
Comment (minudika) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/ApiResponseMessage.java]: Make these variables private

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/SiddhiAppsApi.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import javax.ws.rs.*;
+import javax.ws.rs.core.Response;
+import org.wso2.msf4j.interceptor.annotation.RequestInterceptor;
+
+@RequestInterceptor(AuthenticationInterceptor.class)
+@javax.annotation.Generated(value = "io.swagger.codegen.languages.JavaMSF4JServerCodegen",
+        date = "2018-08-01T11:26:25.925Z")
+public class SiddhiAppsApi implements Microservice {
+    public static final String API_CONTEXT_PATH = "/apis/datasearch";
+    private Logger log = LoggerFactory.getLogger(SiddhiAppsApi.class);
```
Comment (minudika) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/SiddhiAppsApi.java]: Do we need a logger here?

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/findbugs-exclude.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/findbugs-exclude.xml]: Why is he suppression file added? We are not suppressing anything

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            <artifactId>org.wso2.carbon.uiserver</artifactId>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>build-helper-maven-plugin</artifactId>
+                <version>1.9.1</version>
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml]: Move version to parent pom

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            org.osgi.framework.*;version="${org.osgi.framework.version.range}",
+            org.wso2.msf4j.*;version="${msf4j-core.version.range}",
+            org.wso2.siddhi.*;version="${siddhi.version.range}",
+            *;resolution:=optional
+        </import.package>
+        <carbon.component>
+            osgi.service; objectClass="org.wso2.carbon.uiserver.spi.RestApiProvider"
+        </carbon.component>
+        <checkstyle.skip>true</checkstyle.skip>
+        <findbugs.skip>true</findbugs.skip>
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml]: Why is findbugs scan skipped?

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ServiceComponent.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+@Component(
+        name = "org.wso2.carbon.siddhi.apps.api.rest.config.ServiceComponent",
+        service = ServiceComponent.class,
+        immediate = true
+)
+public class ServiceComponent {
+
+    @Activate
+    public void start(){
+
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ServiceComponent.java]: Shouldn't there be permission validated?

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/SiddhiAppsApi.java (hunk 2 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/SiddhiAppsApi.java]: Fixed 9f56a17

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/ApiResponseMessage.java (hunk 1 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/ApiResponseMessage.java]: Fixed a9f5bb3

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ConfigReader.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ */
+package org.wso2.carbon.siddhi.apps.api.rest.config;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConfigReader {
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ConfigReader.java]: Fixed 4fefa54

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        date = "2017-11-01T11:26:25.925Z")
+public class SiddhiAppsApiServiceImpl extends SiddhiAppsApiService {
+
+    private static final Logger log = LoggerFactory.getLogger(SiddhiAppsApiServiceImpl.class);
+    private Gson gson = new Gson();
+
+    @Override
+    public Response getSiddhiAppDefinitions(String id,String appName) {
+
+        String[] hostPort = id.split("_");
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: Fixed 7eeada0

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ConfigReader.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    private  Map<String,Object> configs = readConfigs();
+
+
+    private  static Map<String, Object> readConfigs(){
+        try{
+            return (Map<String,Object>) DataHolder.getInstance()
+                    .getConfigProvider().getConfigurationObject(COMPONENT_NAMESPACE);
+        }catch (Exception e){
+            log.error("Failed to read deplyment.yaml file",e);
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ConfigReader.java]: Fixed bbb8597

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/SiddhiAppsApiService.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+@javax.annotation.Generated(value = "io.swagger.codegen.languages.JavaMSF4JServerCodegen",
+        date = "2017-11-01T11:26:25.925Z")
+public abstract class SiddhiAppsApiService {
+
+    //Get All sidhhi apps from a worker
+    public abstract Response getAllSiddhiApps();
+
+    //Get Siddhi Definitions from a siddhi app
+    public abstract Response getSiddhiAppDefinitions(String id,String appName);
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/gen/java/org/wso2/carbon/siddhi/apps/api/rest/SiddhiAppsApiService.java]: Fixed bfe9ab9

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/worker/WorkerServiceFactory.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.wso2.carbon.siddhi.apps.api.rest.worker;
+
+import org.wso2.carbon.siddhi.apps.api.rest.config.DataHolder;
+
+public class WorkerServiceFactory {
+
+    public static WorkerServiceStub getWorkerHttpsClient(String url, String username, String password){
+        return DataHolder.getInstance().getClientBuilderService().build(username,password,
+                10000,10000,WorkerServiceStub.class,url);
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/worker/WorkerServiceFactory.java]: Fixed e7c64a4

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+                        for (String appName : siddhiAppList) {
+                            if (!siddhiApps.containsKey(appName)) {
+                                siddhiApps.put(appName, workerId);
+                            }
+                        }
+                    }
+                }
+            }catch(feign.RetryableException e){
+                log.warn("Unable to reach the worker "+worker);
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: Fixed e7c64a4

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+            Map<String, AggregationDefinition> aggregationDefinitionMap = siddhiApp.getAggregationDefinitionMap();
+
+            //Get Aggregation defintions with @store annotations
+            for(Map.Entry<String, AggregationDefinition> entry : aggregationDefinitionMap.entrySet()){
+                if(entry.getValue().toString().contains("@store")){
+
+                    //Get Attribute list
+                    StringBuilder attributes  = new StringBuilder(entry.getValue().getAttributeList().get(0).getName()+
+                            " "+entry.getValue().getAttributeList().get(0).getType());
+                    for(int i=1;i<entry.getValue().getAttributeList().size();i++){
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: Fixed 49c5a87

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/bean/SiddhiAppWorker.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Override
+    public String toString() {
+        return "SiddhiAppWorker{" +
+                "appName='" + appName + '\'' +
+                ", worker='" + worker + '\'' +
+                '}';
+    }
+}
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/bean/SiddhiAppWorker.java]: Fixed in bbb8597

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ServiceComponent.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+            unbind = "unregisterConfigProvider"
+    )
+    protected void registerConfigProvider(ConfigProvider configProvider) {
+        DataHolder.getInstance().setConfigProvider(configProvider);
+    }
+
+    protected void unregisterConfigProvider(ConfigProvider configProvider) {
+        DataHolder.getInstance().setConfigProvider(null);
+    }
+
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ServiceComponent.java]: Fixed in bbb8597

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+                        attributes.append(", "+entry.getValue().getAttributeList().get(i).getName()+" "+
+                                entry.getValue().getAttributeList().get(i).getType());
+                    }
+                    SiddhiDefinition siddhiDefinition = new SiddhiDefinition(entry.getKey(),entry.getValue().toString(),"Aggregation",attributes.toString());
```
[the remark this answers is not in the document]
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: Fixed in 49c5a87

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/findbugs-exclude.xml (hunk 3 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/findbugs-exclude.xml]: Fixed in bbb8597

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml (hunk 4 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml]: Fixed in 95e4574

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml (hunk 5 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml]: Fixed in 870c965

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ServiceComponent.java (hunk 6 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/config/ServiceComponent.java]: Implemented permission provider in 870c965

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    <properties>
+        <export.package>
+            org.wso2.carbon.siddhi.apps.api.*
+        </export.package>
+        <import.package>
+            org.osgi.framework.*;version="${org.osgi.framework.version.range}",
+            org.wso2.carbon.analytics.permissions.*; version="${carbon.analytics-common.version.range}",
+            org.wso2.msf4j.*;version="${msf4j-core.version.range}",
+            org.wso2.siddhi.*;version="${siddhi.version.range}",
+            *;resolution:=optional
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml]: If all dependencies are included, then no need "*;"

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+        ArrayList<String> workerList =  cf.getWorkerList(); //list of worker urls
+
+        //Map to store siddhi apps from all the workers
+        Map<String, String> siddhiApps = new ConcurrentHashMap<>();
+
+        workerList.parallelStream().forEach(worker -> {
+            try {
+                feign.Response workerResponse = WorkerServiceFactory.getWorkerHttpsClient(
+                        PROTOCOL + worker, username, password).getSiddhiAppNames();
+                if (workerResponse != null && workerResponse.status() == 200) {
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: Is only Response.Ok from worker run times handled?

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+        ConfigReader cf = new ConfigReader();
+        String username = cf.getUserName();
+        String password = cf.getPassword();
+
+        try {
+            feign.Response workerResponse = WorkerServiceFactory.getWorkerHttpsClient(
+                    PROTOCOL + workerURI, username, password).getSiddhiAppContent(appName);
+
+            //Get App Content
+            SiddhiAppContent siddhiAppContent = gson.fromJson(workerResponse.body().toString(), SiddhiAppContent.class);
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: What if worker returns response that is not 200/null?

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/resources/siddhi-apps-rest-api.yaml [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+  /siddhi-apps:
+    get:
+      summary: Lists Siddhi Applications.
+      description:
+        Provides siddhi apps which contain @store annotated Aggregation, Table, Window definitions.
+      produces:
+      - application/json
+      responses:
+        '200':
+          description: The Siddhi Applications are successfully retrieved.
```
Comment (Niveathika92) [inline components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/resources/siddhi-apps-rest-api.yaml]: What is the schema of the response?

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml (hunk 17 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/pom.xml]: Fixed in 316b47d

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java (hunk 18 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: Fixed in a8ae221

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java (hunk 19 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/java/org/wso2/carbon/siddhi/apps/api/rest/impl/SiddhiAppsApiServiceImpl.java]: Fixed in a8ae221

File: components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/resources/siddhi-apps-rest-api.yaml (hunk 20 above)
Comment (LakshanSS) [reply components/dashboards/org.wso2.carbon.dashboards.siddhi.apps.api/src/main/resources/siddhi-apps-rest-api.yaml]: Fixed in a8ae221

**Comment by Niveathika92** [comment]
@LakshanSS, As discussed offline, shall we have a code review before merging these?

**Pull request closed without merging**